// @sentry/node/cjs/integrations/tracing/graphql.js
Object.defineProperty(exports, '__esModule', { value: true });

const instrumentationGraphql = require('@opentelemetry/instrumentation-graphql');
const core = require('@sentry/core');
const opentelemetry = require('@sentry/opentelemetry');
const addOriginToSpan = require('../../utils/addOriginToSpan.js');

const _graphqlIntegration = ((_options = {}) => {
  const options = {
    ignoreResolveSpans: true,
    ignoreTrivialResolveSpans: true,
    ..._options,
  };

  return {
    name: 'Graphql',
    setupOnce() {
      opentelemetry.addOpenTelemetryInstrumentation(
        new instrumentationGraphql.GraphQLInstrumentation({
          ...options,
          responseHook(span) {
            addOriginToSpan.addOriginToSpan(span, 'auto.graphql.otel.graphql');
          },
        }),
      );
    },
  };
}) ;

/**
 * GraphQL integration
 *
 * Capture tracing data for GraphQL.
 */
const graphqlIntegration = core.defineIntegration(_graphqlIntegration);

exports.graphqlIntegration = graphqlIntegration;
